linux/devdoc/socket_transport_linux.md

```diff
@@ -0,0 +1,284 @@
+```
+
+`socket_transport_destroy` destroys all data stored in the `SOCKET_TRANSPORT_HANDLE` object.
+
+If `socket_transport` is `NULL` `socket_transport_destroy` shall do nothing.
+
+`socket_transport_destroy` shall call `sm_destroy` to destroy the sm object.
+
+`socket_transport_destroy` shall free the `SOCKET_TRANSPORT_HANDLE` object.
+
+### socket_transport_connect
+
+```c
+MOCKABLE_FUNCTION(, int, socket_transport_connect, SOCKET_TRANSPORT_HANDLE, socket_transport, const char*, hostname, uint16_t, port, uint32_t, connection_timeout_ms);
+```
+
+`socket_transport_connect` shall connect to a specified endpoint.
+
+If `socket_transport` is `NULL`, `socket_transport_connect` shall fail and return a non-zero value.
+
+If `hostname` is `NULL`, `socket_transport_connect` shall fail and return a non-zero value.
+
+If `port` is `0`, `socket_transport_connect` shall fail and return a non-zero value.
+
+If the transport type is not `SOCKET_CLIENT`, `socket_transport_connect` shall fail and return a non-zero value.
+
+`socket_transport_connect` shall call `sm_open_begin` to begin the open.
+
+If `sm_open_begin` does not return `SM_EXEC_GRANTED`, `socket_transport_connect` shall fail and return a non-zero value.
+
+`socket_transport_connect` shall call `socket` with the params `AF_INET`, `SOCK_STREAM` and `0`.
+
+`socket_transport_connect` shall call `connect` to connect to the endpoint.
+
+`socket_transport_connect` shall set the socket to non-blocking.
+
+If successful `socket_transport_connect` shall call `sm_open_end` with `true`.
+
+If any failure is encountered, `socket_transport_connect` shall call `sm_open_end` with `false`, fail and return a non-zero value.
+
+### socket_transport_disconnect
+
+```c
+MOCKABLE_FUNCTION(, void, socket_transport_disconnect, SOCKET_TRANSPORT_HANDLE, socket_transport);
+```
+
+`socket_transport_disconnect` shall disconnect a connected socked from its endpoint.
+
+If `socket_transport` is `NULL`, `socket_transport_disconnect` shall fail and return.
+
+`socket_transport_disconnect` shall call `sm_close_begin` to begin the closing process.
+
+If `sm_close_begin` does not return `SM_EXEC_GRANTED`, `socket_transport_disconnect` shall fail and return.
+
+`socket_transport_disconnect` shall call `close` to disconnect the connected socket.
+
+`socket_transport_disconnect` shall call `sm_close_end`.
+
+### socket_transport_send
+
+```c
+MOCKABLE_FUNCTION(, SOCKET_SEND_RESULT, socket_transport_send, SOCKET_TRANSPORT_HANDLE, socket_transport, SOCKET_BUFFER*, payload, uint32_t, buffer_count, uint32_t*, bytes_written, uint32_t, flags, void*, overlapped_data);
+```
+
+If `socket_transport` is `NULL`, `socket_transport_send` shall fail and return `SOCKET_SEND_ERROR`.
+
+If `payload` is `NULL`, `socket_transport_send` shall fail and return `SOCKET_SEND_ERROR`.
+
+If `buffer_count` is `0`, `socket_transport_send` shall fail and return `SOCKET_SEND_ERROR`.
+
+`socket_transport_send` shall call `sm_exec_begin`.
```

Comment:
>sm_exec_begin [](http://example.com/codeflow?start=36&length=13)

what's the threading here? What happens if 2 socket_transport_send are called at the same time? Will the 2 threads compete over the `send` to the same socket?

Reply:
only asking because if only 1 can execute at the same time then it is sm_barrier...

Reply:
hmm, that's a good question that will require some research.  I believe sockets are thread-safe, but I shall have to either write a test or look at more documentation.

Reply:
For now I believe it's OK for multiple threads to send on a sockets.